### PR TITLE
Support for Symfony 2.8+ autowiring

### PR DESCRIPTION
**Pull request by lemoinem**
Here is the PR proposed in #253 

I also added a dev dependency on symfony/expression-language because the functional tests were crashing otherwise.

I updated the tests to include testing for the support of autowiring, as much as I could understand the current structure of the tests. Tell me if you see any test case I should add.

I took the option to throw an exception if we use "autowire" with a version of Symfony that doesn't support it. However, autowiringType is silently ignored if there is no support for it.

### CODE REVIEW

**Comment by lemoinem** [comment]
@Ener-Getick @ksom here is the PR from #253 If you'd like to review it!

File: Tests/Metadata/Driver/Fixture/Service.php [hunk 1]
```diff
@@ -11,7 +11,9 @@
  *     decorates="test.service",
  *     decoration_inner_name="original.test.service",
  *     deprecated="use new.test.service instead",
- *     public=false
+ *     public=false,
+ *     autowire=false,
+ *     autowiring_types={"JMS\DiExtraBundle\Tests\Metadata\Driver\Fixture\Service"},
```
Comment (Ener-Getick) [inline Tests/Metadata/Driver/Fixture/Service.php]: I think you can use the syntax ``Service::class`` here

File: Metadata/MetadataConverter.php [hunk 2]
```diff
@@ -63,7 +63,22 @@ public function convert(ClassHierarchyMetadata $metadata)
             if (null !== $classMetadata->arguments) {
                 $definition->setArguments($classMetadata->arguments);
             }
+            if (null !== $classMetadata->autowire) {
+                if (!method_exists($definition, 'setAutowire')) {
+                    throw new InvalidAnnotationException(
+                        sprintf(
+                            "autowiring require symfony >=2.8 on class %s"
```
Comment (Ener-Getick) [inline Metadata/MetadataConverter.php]: ``requires``

Comment (Ener-Getick) [inline Metadata/MetadataConverter.php]: I would reword this a bit in fact ``You must use symfony 2.8 or higher to use autowiring on the class %s.`` what do you think?

File: Annotation/Service.php [hunk 3]
```diff
@@ -53,4 +53,10 @@
 
     /** @var array<string> */
     public $environments = array();
+
+    /** @var boolean */
+    public $autowire;
+
+    /** @var array<string> */
+    public $autowiring_types = array();
```
Comment (Ener-Getick) [inline Annotation/Service.php]: Should be ``autowiringTypes``

File: Metadata/MetadataConverter.php (hunk 2 above)
Comment (Ener-Getick) [inline Metadata/MetadataConverter.php]: And can be inlined.

File: composer.json [hunk 4]
```diff
@@ -29,6 +29,7 @@
         "symfony/yaml": "~2.3|~3.0",
         "symfony/browser-kit": "~2.3|~3.0",
         "symfony/security-bundle": "~2.3",
+        "symfony/expression-language": "~2.6|~3.0.0",
```
Comment (Ener-Getick) [inline composer.json]: Should be ``~2.6|~3.0``

**Comment by Ener-Getick** [comment]
Except my comments, looks good to me :+1:

File: Metadata/MetadataConverter.php (hunk 2 above)
Comment (lemoinem) [inline Metadata/MetadataConverter.php]: Sounds good!
I based it on https://github.com/schmittjoh/JMSDiExtraBundle/pull/254/files#diff-795eb25e349a10927c913d862c5cf87aL73 I will fix this other one too.

File: Annotation/Service.php (hunk 3 above)
Comment (lemoinem) [inline Annotation/Service.php]: The point was to keep the snake_case syntax for annotations' attributes. Just like `decoration_inner_name` (https://github.com/schmittjoh/JMSDiExtraBundle/pull/254/files#diff-2c2011597fae2a64c228c5d101095beaL49)

Which one should it be? Switch both to camelCase or keep snake_case for both?

Comment (Ener-Getick) [inline Annotation/Service.php]: I'd say camel case and we should deprecate any attributes snack cased. 
If I'm not wrong all majors libraries using annotations use camel case right?

Comment (lemoinem) [inline Annotation/Service.php]: Indeed, I'll switch this one and prepare an additional PR with the rename/deprecation of `decoration_inner_name` (it seems important enough to warrant its own PR.)

File: Tests/Metadata/Driver/Fixture/Service.php (hunk 1 above)
Comment (lemoinem) [inline Tests/Metadata/Driver/Fixture/Service.php]: `Service::class` makes the travis run for PHP 5.3 + prefer-lowest fail...
However PHP 5.3 with standard dependencies pass...

Still, I guess I will revert to the literal. (Unless you know what (dev) dependency I should update to make it pass?)

Comment (Ener-Getick) [inline Tests/Metadata/Driver/Fixture/Service.php]: I don't know :-\ let's just revert this that's not important.

**Comment by lemoinem** [comment]
@Ener-Getick: Ok, all the tests are :green_heart:! Thanks for the review and real fast feedback (:
I took care of all your comments.

I'll prepare the PR to rename `decoration_inner_name` in the next few days.

**Comment by Ener-Getick** [comment]
Great thank you for this contribution ! ☺
